File: packaging/smoke_sandbox.mjs
/** Exercise every bundled MCP tool from inside the disposable sandbox. */
import { spawn } from 'node:child_process';
import fs from 'node:fs';
import { createInterface } from 'node:readline';

const ENTRYPOINT = '/opt/lumen/mcp/computer-use/dist/index.js';
const WORKSPACE = '/tmp/lumen-mcp-smoke';
const EXPECTED_TOOLS = ['bash_tool', 'create_file', 'str_replace', 'view'];

let nextId = 1;
const pending = new Map();
const server = spawn('node', [ENTRYPOINT], {
  env: { ...process.env, TRANSPORT: 'stdio' },
  stdio: ['pipe', 'pipe', 'pipe'],
});
let stderr = '';
server.stderr.on('data', chunk => { stderr += chunk.toString(); });

const lines = createInterface({ input: server.stdout });
lines.on('line', line => {
  let message;
  try {
    message = JSON.parse(line);
  } catch {
    return;
  }
  const waiter = pending.get(message.id);
  if (waiter) {
    pending.delete(message.id);
    clearTimeout(waiter.timer);
    waiter.resolve(message);
  }
});

function send(method, params = {}) {
  return new Promise((resolve, reject) => {
    const id = nextId++;
    const timer = setTimeout(() => {
      pending.delete(id);
      reject(new Error(`MCP request timed out: ${method}\n${stderr}`));
    }, 15_000);
    pending.set(id, { resolve, reject, timer });
    server.stdin.write(`${JSON.stringify({ jsonrpc: '2.0', id, method, params })}\n`);
  });
}

async function callTool(name, args) {
  const response = await send('tools/call', { name, arguments: args });
  if (response.error) throw new Error(`${name}: ${JSON.stringify(response.error)}`);
  const text = (response.result.content || []).map(item => item.text || '').join('');
  if (response.result.isError) throw new Error(`${name}: ${text}`);
  return text;
}

async function main() {
  fs.rmSync(WORKSPACE, { recursive: true, force: true });
  fs.mkdirSync(WORKSPACE, { recursive: true });

  await send('initialize', {
    protocolVersion: '2024-11-05',
    capabilities: {},
    clientInfo: { name: 'lumen-sandbox-smoke', version: '1.0' },
  });

  const listed = await send('tools/list');
  const tools = listed.result.tools.map(tool => tool.name).sort();
  if (JSON.stringify(tools) !== JSON.stringify(EXPECTED_TOOLS)) {
    throw new Error(`unexpected tools: ${JSON.stringify(tools)}`);
  }

  const path = `${WORKSPACE}/smoke.txt`;
  await callTool('create_file', {
    description: 'sandbox smoke test', path, file_text: 'alpha beta\n',
  });
  await callTool('str_replace', {
    description: 'sandbox smoke test', path, old_str: 'alpha', new_str: 'lumen',
  });
  const viewed = await callTool('view', {
    description: 'sandbox smoke test', path,
  });
  if (!viewed.includes('lumen beta')) throw new Error(`unexpected view output: ${viewed}`);

  const bashText = await callTool('bash_tool', {
    description: 'sandbox smoke test',
    command: `test -f ${path} && printf mcp-ok`,
  });
  const bashResult = JSON.parse(bashText);
  if (bashResult.returncode !== 0 || !bashResult.stdout.includes('mcp-ok')) {
    throw new Error(`unexpected bash output: ${bashText}`);
  }

  console.log(JSON.stringify({ tools, workspace_write: true, bash: true }));
}

try {
  await main();
} finally {
  server.kill('SIGTERM');
  lines.close();
}

File: packaging/smoke_sandbox.py
#!/usr/bin/env python3
"""Exercise the bundled computer-use MCP server through a Docker container."""
from __future__ import annotations

import argparse
import asyncio
import json
import os
import sys
import tempfile
import uuid
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

import docker_cli  # noqa: E402
from mcp import ClientSession, StdioServerParameters  # noqa: E402
from mcp.client.stdio import stdio_client  # noqa: E402


MCP_ENTRYPOINT = "/opt/lumen/mcp/computer-use/dist/index.js"
EXPECTED_TOOLS = {"view", "create_file", "str_replace", "bash_tool"}


def _docker(args: list[str], *, check: bool = True):
    completed = docker_cli.run(args)
    if check and completed.returncode != 0:
        detail = (completed.stderr or completed.stdout or "Docker command failed").strip()
        raise RuntimeError(f"docker {' '.join(args[:2])} failed: {detail}")
    return completed


def _text(result: object) -> str:
    return "".join(
        getattr(item, "text", "")
        for item in getattr(result, "content", [])
        if getattr(item, "type", "") == "text"
    )


def _assert_ok(result: object, label: str) -> str:
    text = _text(result)
    if getattr(result, "isError", False):
        raise RuntimeError(f"{label} failed: {text}")
    return text


async def _exercise_server(container_name: str) -> dict[str, object]:
    executable = docker_cli.executable()
    params = StdioServerParameters(
        command=executable,
        args=[
            "exec",
            "-i",
            "--workdir",
            "/workspace",
            container_name,
            "node",
            MCP_ENTRYPOINT,
        ],
        env=dict(os.environ),
    )

    async with stdio_client(params) as (reader, writer):
        async with ClientSession(reader, writer) as session:
            await session.initialize()
            discovered = {tool.name for tool in (await session.list_tools()).tools}
            if discovered != EXPECTED_TOOLS:
                raise RuntimeError(
                    f"unexpected tools: expected {sorted(EXPECTED_TOOLS)}, got {sorted(discovered)}"
                )

            created = await session.call_tool(
                "create_file",
                {
                    "description": "sandbox smoke test",
                    "path": "/workspace/lumen-smoke.txt",
                    "file_text": "alpha beta\n",
                },
            )
            _assert_ok(created, "create_file")

            replaced = await session.call_tool(
                "str_replace",
                {
                    "description": "sandbox smoke test",
                    "path": "/workspace/lumen-smoke.txt",
                    "old_str": "alpha",
                    "new_str": "lumen",
                },
            )
            _assert_ok(replaced, "str_replace")

            viewed = await session.call_tool(
                "view",
                {
                    "description": "sandbox smoke test",
                    "path": "/workspace/lumen-smoke.txt",
                },
            )
            viewed_text = _assert_ok(viewed, "view")
            if "lumen beta" not in viewed_text:
                raise RuntimeError(f"view returned unexpected content: {viewed_text}")

            bashed = await session.call_tool(
                "bash_tool",
                {
                    "description": "sandbox smoke test",
                    "command": "pwd && test -f /workspace/lumen-smoke.txt && printf mcp-ok",
                },
            )
            bash_payload = json.loads(_assert_ok(bashed, "bash_tool"))
            if bash_payload.get("returncode") != 0 or "mcp-ok" not in bash_payload.get("stdout", ""):
                raise RuntimeError(f"bash_tool returned unexpected result: {bash_payload}")

    return {"tools": sorted(discovered), "workspace_write": True, "bash": True}


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--image", default="lumen-sandbox")
    args = parser.parse_args()

    container_name = f"lumen-mcp-smoke-{uuid.uuid4().hex[:12]}"
    with tempfile.TemporaryDirectory(prefix="lumen-mcp-smoke-") as workspace:
        try:
            _docker(
                [
                    "run",
                    "--detach",
                    "--name",
                    container_name,
                    "--label",
                    "com.lumen.smoke=true",
                    "--volume",
                    f"{workspace}:/workspace",
                    args.image,
                ]
            )
            _docker(
                ["exec", container_name, "test", "-f", MCP_ENTRYPOINT]
            )
            result = asyncio.run(_exercise_server(container_name))
            result.update({"image": args.image, "entrypoint": MCP_ENTRYPOINT})
            print(json.dumps(result, indent=2))
        finally:
            _docker(["rm", "--force", container_name], check=False)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
